### PR TITLE
XYZwriter also for moleculekit object

### PR DESCRIPTION
**Pull request by AntonioMirarchi**
I notices that the moleculekit object give an error during the xyz_writing.  Here I propose an update version that it's able to work with both npzmol and moleculekit object. 
Key points:
- [x] `element` attribute to the npzMolecule object using moleculekit.periodictable so we have the same attribute as in moleculekit object;
- [x] new name for the `xyz_writer` function with relative documentation;
- [x] now the xyz_writer function use directly the `mol.element` to write the element symbol in the xyz file;

The error which I'm referring to fix: 
image

### CODE REVIEW

**Comment by AntonioMirarchi** [comment]
I fixed also a warning from torchmd/systems.py:49: 
`UserWarning`: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).

**Comment by AntonioMirarchi** [comment]
@stefdoerr could you check this please? 
At the moment the xyz writer is not working with the moleculekit object.

**Review by RaulPPelaez — APPROVED** [review]

**Comment by AntonioMirarchi** [comment]
Because when I created it, I was working only with `npzMol class` and for this reason I didn't test with moleculekit object.  When I use it with the moleculekit object, `mol.z` give me this error:
image

Now using directly `mol.element`, it works fine with both object. Also for this I changed the npzMol attributes according to the moleculekit object attributes.

**Comment by stefdoerr** [comment]
image
what's wrong with this file?

**Comment by AntonioMirarchi** [comment]
Ok if there is no problem then you can close this PR. I created it just because I got that error using moleculekit object from pdb/psf file.

**Comment by AntonioMirarchi** [comment]
I wrote this code if you want to reproduce the error that I'm getting, and get why we need to change the xyz_writer function.
```python 
import numpy as np
from moleculekit.molecule import Molecule

# function of this PR 
def xyz_writer(input_file, output_file, mol_elements):
    npy_file = np.load(input_file)
    Nsteps = npy_file.shape[2]
    Nats = npy_file.shape[0]
    for i in range(Nsteps):
        with open(output_file, "a") as f:
            if "forces" not in input_file:
                f.write(str(Nats))
            f.write("\n\n")
            for j in range(Nats):
                if "forces" not in input_file:
                    f.write(mol_elements[j])
                    f.write(" ")
                for k in range(3):
                    f.write(str(npy_file[j, k, i]))
                    f.write(" ")
                f.write("\n")

# function from torchmd repo 
def converter_xyz_output(input_file, output_file, z=None):
    from moleculekit.periodictable import periodictable_by_number
    mol_elements = np.array(z)
    npy_file = np.load(input_file)
    Nsteps = npy_file.shape[2]
    Nats = npy_file.shape[0]
    for i in range(Nsteps):
        with open(output_file, "a") as f:
            if "forces" not in input_file:
                f.write(str(Nats))
            f.write("\n\n")
            for j in range(Nats):
                if "forces" not in input_file:
                    f.write(periodictable_by_number[mol_elements[j]].symbol)
                    f.write(" ")
                for k in range(3):
                    f.write(str(npy_file[j, k, i]))
                    f.write(" ")
                f.write("\n")


if __name__ == "__main__":
    npy_traj_path = "traj.npy"
    xyz_traj_path = "traj.xyz"

    mol = Molecule("3PTB", name="Trypsin")
    print(f"test on : Trypsin")
    print(f"mol coords: {mol.coords.shape}")

    # generate a random trajectory of 10 frames
    npy_traj = np.repeat(mol.coords, 10, axis=2)
    np.save("traj.npy", npy_traj)
    print(f"npy_traj of shape {npy_traj.shape} successfully created!")

    try:
        converter_xyz_output(npy_traj_path, xyz_traj_path, mol.z)
        print("Success to use converter_xyz_output()")
    except:
        print("Fail to use converter_xyz_output()")

    try:
        xyz_writer(npy_traj_path, xyz_traj_path, mol.element)
        print("Success to use xyz_writer()")
    except:
        print("Fail to use xyz_writer()")
```

**Comment by stefdoerr** [comment]
Fist of all `mol.z` is not the element. It's the Z coordinate of the first frame of the Molecule for every atom
```
In [3]: mol.z
Out[3]: array([20.309, 19.112, 18.287, ..., 28.43 , 28.215, 27.945], dtype=float32)

In [4]: mol.y
Out[4]: array([ 9.599,  8.707,  8.966, ..., 20.922, 18.227, 20.916], dtype=float32)

In [5]: mol.x
Out[5]: array([-8.096, -8.084, -9.354, ...,  5.803, 17.331, 17.388], dtype=float32)
```

Second, this one also works. Although it will not write forces if that's what you want. But that's just an non-standard xyz format anyway
```
mol.write('traj.xyz')
```

This works fine as well
```py
def converter_xyz_output(input_file, output_file, z=None):
    from moleculekit.periodictable import periodictable

    mol_elements = np.array(z)
    npy_file = np.load(input_file)
    Nsteps = npy_file.shape[2]
    Nats = npy_file.shape[0]
    for i in range(Nsteps):
        with open(output_file, "a") as f:
            if "forces" not in input_file:
                f.write(str(Nats))
            f.write("\n\n")
            for j in range(Nats):
                if "forces" not in input_file:
                    f.write(periodictable[mol_elements[j]].symbol)
                    f.write(" ")
                for k in range(3):
                    f.write(str(npy_file[j, k, i]))
                    f.write(" ")
                f.write("\n")

converter_xyz_output(npy_traj_path, xyz_traj_path, mol.element)
```

**Comment by stefdoerr** [comment]
If you want to use the existing `converter_xyz_output` function then just do the following to get a "z-array"

```py
from moleculekit.periodictable import periodictable
z = np.array([periodictable[el].number for el in mol.element])
converter_xyz_output(npy_traj_path, xyz_traj_path, z)
```

File: torchmd/run.py [hunk 1]
```diff
@@ -277,7 +277,7 @@ def dynamics(args, mol, system, forces):
     for k in range(args.replicas):
         npy_name = os.path.join(args.log_dir, args.output + f"_{k}.npy")
         xyz_name = os.path.join(args.log_dir, args.output + f"_{k}.xyz")
-        converter_xyz_output(npy_name, xyz_name, mol.z)
```
Comment (AntonioMirarchi) [inline torchmd/run.py:280]: So here we are using `mol.z` and as you said are z coordinates for a moleculekit.Molecule.object, so it's going to fail

**Comment by AntonioMirarchi** [comment]
> Second, this one also works. Although it will not write forces if that's what you want. But that's just an non-standard xyz format anyway
> 
> ```
> mol.write('traj.xyz')
> ```
> 
> This works fine as well
> 
> ```python
> def converter_xyz_output(input_file, output_file, element):
>     from moleculekit.periodictable import periodictable
> 
>     mol_elements = np.array(z)
>     npy_file = np.load(input_file)
>     Nsteps = npy_file.shape[2]
>     Nats = npy_file.shape[0]
>     for i in range(Nsteps):
>         with open(output_file, "a") as f:
>             if "forces" not in input_file:
>                 f.write(str(Nats))
>             f.write("\n\n")
>             for j in range(Nats):
>                 if "forces" not in input_file:
>                     f.write(element[j])
>                     f.write(" ")
>                 for k in range(3):
>                     f.write(str(npy_file[j, k, i]))
>                     f.write(" ")
>                 f.write("\n")
> 
> converter_xyz_output(npy_traj_path, xyz_traj_path, mol.element)
> ```

- we cannot use `mol.write(traj.xyz)` since the npzMol class just pretend to be similar to a molecule.object but it's not; 
- that function is broken, you are calling z inside but it's not an initial argument

**Comment by stefdoerr** [comment]
```py
def converter_xyz_output(input_file, output_file, element):
    npy_file = np.load(input_file)
    Nsteps = npy_file.shape[2]
    Nats = npy_file.shape[0]
    for i in range(Nsteps):
        with open(output_file, "a") as f:
            if "forces" not in input_file:
                f.write(str(Nats))
            f.write("\n\n")
            for j in range(Nats):
                if "forces" not in input_file:
                    f.write(element[j])
                    f.write(" ")
                for k in range(3):
                    f.write(str(npy_file[j, k, i]))
                    f.write(" ")
                f.write("\n")

converter_xyz_output(npy_traj_path, xyz_traj_path, mol.element)
```
you get the idea :P but as I said if you don't want to modify that function just create the z array from the mol.element as above

File: torchmd/run.py (hunk 1 above)
Comment (stefdoerr) [reply torchmd/run.py:280]: I don't know who wrote that :/ it's obviously wrong

Comment (AntonioMirarchi) [reply torchmd/run.py:280]: I wrote that when I was using only npzMol class. Now that I'm using also moleculekit object, I know the error and I'm trying to fix it. Also the npzMol class is getting improved in this PR.

**Comment by AntonioMirarchi** [comment]
About the "forces" in the `xyz_writer()` we can remove it. Just less code, also because I think the forces are never saved in the npy traj file.

File: torchmd/run.py [hunk 2]
```diff
@@ -286,7 +286,7 @@ def dynamics(args, mol, system, forces):
...
+        xyz_writer(npy_name, xyz_name, mol.element)
```
Comment (stefdoerr) [inline torchmd/run.py:289]:
```py
npy_traj = np.load(npy_name)
molc = mol.copy()
molc.coords = npy_traj.astype(np.float32)
molc.write(xyz_file)
```

No need for a custom writer then I assume

Comment (AntonioMirarchi) [reply torchmd/run.py:289]: it's not going to work if you have a npzMolecule object; there is no copy and write function

**Comment by AntonioMirarchi** [comment]
shall we merge this?

File: torchmd/npzMol.py [hunk 3]
```diff
@@ -1,33 +1,28 @@
...
-        }
+        """Class to load npz files in torchmd, essentially a wrapper around the npz file. Necessary information is extracted from the npz file
+        and stored in the class, such as the coordinates and the atomic numbers. Additional properties such as the box, the charges and the bonds
+        are also stored in the class but are not necessary for the simulation."""
+
         self.data = np.load(file)
-        self.z = torch.from_numpy(self.data["z"])
-        self.coords = torch.from_numpy(self.data["coord"])[:, :, None]
+        self.z = np.array(self.data["z"])
+        self.coords = np.array(self.data["coord"])[:, :, None]
```
Comment (stefdoerr) [inline torchmd/npzMol.py:14]: I don't understand why here you stopped converting to torch tensor and just make numpy array and then `masses` two lines down is converted to torch tensor. Shouldn't they all be either numpy arrays or torch tensors?

Comment (AntonioMirarchi) [reply torchmd/npzMol.py:14]: The npz mol class is building a mol object similar to the one from moleculekit.molecule so in this case we don't need the tensor but just numpy array that will transformed later to tensor. 

About the masses: the original idea from which this class came out was to run torchmd using only `pos` and `z`, no ff object and so this class returns directly a tensor for the masses because then there is this  [if](https://github.com/torchmd/torchmd/blob/1bad86ec08cec101320d66b24529a032ace4d016/torchmd/parameters.py#L116) that let you to skip make_masses() since it needs ff. Maybe we can remove it or put it as np.array, since now we are considering to use the  mol.element #61

Comment (stefdoerr) [reply torchmd/npzMol.py:14]: Also using elements instead of FF masses is not good because they can be different and will affect the results of the simulations.

Comment (AntonioMirarchi) [reply torchmd/npzMol.py:14]: I wrote that `if` probably when I wrote the npzMol class.

**Review by stefdoerr — APPROVED** [review]
looks ok

**Comment by AntonioMirarchi** [comment]
I changed all the npzMolecule  roperties to numpy.array  and the `if isinstance()` for the masses. I think that at this point we can also remove that `if` and update `make_masses()` to deal with all the possible conditions but let's do it in #61
